kafka.go

```diff
@@ -14,148 +14,296 @@
-		configAsync = configureKafka(producerConfiguration[0])
-		configSync = configureKafka(producerConfiguration[1])
-	case len(producerConfiguration) == 1:
-		configAsync = configureKafka(producerConfiguration[0])
-		configSync = sarama.NewConfig()
-	default:
-		configAsync = sarama.NewConfig()
-		configSync = sarama.NewConfig()
-	}
-
-	asyncProducer, err := sarama.NewAsyncProducer(brokerList, configAsync)
-	if err != nil {
-		return nil, err
+func setConfig(writerConfig *kafka.WriterConfig, readerConfig *kafka.ReaderConfig, config *KafkaConfig) {
+
+	if config.ReadTimeout != 0 {
+		writerConfig.ReadTimeout = config.WriteTimeout
+	}
+	if config.WriteTimeout != 0 {
+		writerConfig.WriteTimeout = config.WriteTimeout
 	}
+	if config.DialTimeout != 0 {
+		dialer := &kafka.Dialer{
+			Timeout: config.DialTimeout,
+		}
+		writerConfig.Dialer = dialer
+		readerConfig.Dialer = dialer
+	}
+}
 
-	configSync.Producer.Return.Successes = true
-	syncProducer, err := sarama.NewSyncProducer(brokerList, configSync)
-	if err != nil {
-		return nil, err
+// NewKafkaClient create a new instance of KafkaClient
+func NewKafkaClient(brokers []string, prefix string, config ...*KafkaConfig) (*KafkaClient, error) {
+
+	logrus.Debug("create new kafka client")
+
+	writerConfig := &kafka.WriterConfig{
+		Brokers:  brokers,
+		Balancer: &kafka.LeastBytes{},
+	}
+
+	readerConfig := &kafka.ReaderConfig{
+		Brokers:  brokers,
+		MaxBytes: defaultReaderSize,
 	}
 
-	go listenToError(asyncProducer)
+	// set client configuration
+	// only uses first KafkaConfig arguments
+	if len(config) > 0 {
+		setConfig(writerConfig, readerConfig, config[0])
+	}
 
 	return &KafkaClient{
-		realm:         realm,
-		asyncProducer: asyncProducer,
-		syncProducer:  syncProducer,
+		prefix:          prefix,
+		publishConfig:   *writerConfig,
+		subscribeConfig: *readerConfig,
+		subscribeMap:    &sync.Map{},
 	}, nil
 }
 
-func listenToError(producer sarama.AsyncProducer) {
-	for err := range producer.Errors() {
-		logrus.Error("unable to push event to Kafka stream: ", err)
+// Publish send event to single or multiple topic with exponential backoff retry
+func (client *KafkaClient) Publish(publishBuilder *PublishBuilder) {
+	logrus.Debugf("publish event %s into topic %s", publishBuilder.eventName, publishBuilder.topic)
+
+	if publishBuilder == nil {
+		logrus.Error("unable to publish nil event")
+		return
+	}
+	message, err := constructEvent(publishBuilder)
+	if err != nil {
+		logrus.Errorf("unable to construct event : %s , error : %v", publishBuilder.eventName, err)
+		return
+	}
+
+	config := client.publishConfig
+	for _, pubTopic := range publishBuilder.topic {
+		topic := pubTopic
+		go func() {
+			err = backoff.RetryNotify(func() error {
+				return client.publishEvent(publishBuilder.ctx, topic, config, message)
+			}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxBackOffCount),
+				func(_ error, _ time.Duration) {
+					logrus.Debugf("retrying publish event: error %v : ", err)
```

Comment:
you'd need more info to debug this, e.g. topic, message, etc

Reply:
sure, will update it
thanks